src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -2571,5 +4014,129 @@ async Unknown local()
+using System;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+
+class Test
+{
+    static async Task Main()
+    {
+        await foreach (string i in new Derived().Iter())
+        {
+            Console.Write(i);
+        }
+    }
+}
+
+class Base
+{
+    public virtual string Func()
+    {
+        return ""Base.Func;"";
+    }
+}
+
+class Derived: Base
+{
+    public override string Func()
+    {
+        return ""Derived.Func;"";
+    }
+
+    public async IAsyncEnumerable<string> Iter()
+    {
+        await Task.CompletedTask;
+        yield return base.Func();
+        yield return this.Func();
+    }
+}";
+            var comp = CreateCompilationWithAsyncIterator(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "Base.Func;Derived.Func;");
+        }
+
+        [Fact]
+        public void TestIteratorWithBaseAccessInLambda()
+        {
+            // modified version of corresponding CodeGenIterators test
+            var source = @"
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+static class M1
+{
+    class B1<T>
+    {
+        public virtual string F<U>(T t, U u)
+        {
+            return ""B1::F;"";
+        }
+    }
+
+    class Outer<V>
+    {
+        public class B2 : B1<V>
+        {
+            public override string F<U>(V t, U u)
+            {
+                return ""B2::F;"";
+            }
+
+            public async Task Test()
+            {
+                Func<Task> m = async () =>
+                    {
+                        await foreach (string i in this.Iter())
+                        {
+                            Console.Write(i);
+                        }
+                    };
+                await m();
+            }
+
+            public async IAsyncEnumerable<string> Iter()
+            {
+                V v = default(V);
+                int i = 0;
+                string s = null;
+
+                Func<string> f = () => { Func<Func<V, int, string>> ff = () => base.F<int>; return ff()(v, i); };
+                yield return f();
+
+                f = () => { Func<Func<V, string, string>> ff = () => this.F<string>; return ff()(v, s); };
+                yield return f();
+
+                f = () => { Func<Func<V, int, string>> ff = () => { i++; return base.F<int>; }; return ff()(v, i); };
+                yield return f();
+                await Task.CompletedTask;
+            }
+        }
+    }
+
+    class D<X> : Outer<X>.B2
+    {
+        public override string F<U>(X t, U u)
+        {
+            return ""D::F;"";
+        }
+    }
+
+    static async Task Main()
+    {
+        await (new D<int>()).Test();
+    }
+}";
+            var comp = CreateCompilationWithAsyncIterator(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "B1::F;D::F;B1::F;");
```

Comment:
I think it would be good to have tests that uses `goto` to jump out of the `try` block. Especially one that re-enters, like

```C#

public class C {
    public IEnumerable M() {
        start:
        try{
            yield return 0;
            goto start;
        } 
        finally
        {
        }
    }
}